### DISCUSSION TITLE
Passing states between routes

### DISCUSSION THREAD

**Discussion by yahelcohen01** [TanStack/router#1186] [Q&A] [answered]
I have a state that I want to use its setState in two diffrent routes.

**Comment by SeanCassiere** [comment, marked as answer]
Edit: [This discussion here](https://github.com/TanStack/router/discussions/1342#discussioncomment-8848490) details the process of passing state between routes. 

> 🚧 The content below, can now be considered to be outdated, but the discussion matter may be useful for understanding why this `state` field is only validated at the type-level and not at runtime.

If you want to pass state between Routes, the search params or query params are your go to option.

If you want your data to be passed outside of the Router you have the built-in React-context pathway or third-party solutions like Jotai and Zustand.

What is this data you need to pass between routes? Often times I've found that search-params are the best solution.

**Reply by yahelcohen01** [reply]
I need to pass some setState to other component. It seems like Zustand is the option. I'll create a store which contains the state

**Reply by IanVS** [reply]
I've found it useful in react-router to occasionally place some state into the history, such as a `from` path.  This way, when I navigate to a particular part of my app I can set the state.from, then later on when the user closes that UI, I can return the user to their original location.  

Or, another use I've had for it before, is determining whether to open a screen in a modal or on a page of its own.  When I click a link I can set some state on the history, then if the component sees that, it opens in a modal.  But if the page is refreshed the state isn't set and it opens full-page.  

It sounds like there's no built-in way to do this within tanstack-router, and we should be using some other app-wide state instead?

**Reply by SeanCassiere** [reply]
@IanVS yes.
As mentioned [here](https://discord.com/channels/719702312431386674/1209077879682633738), currently Tanstack Router doesn't support the `state` property during `router.navigate` because of its inconsistent behaviour in how the type-safety can be guaranteed between navigation.

**Reply by nyngwang** [reply]
@SeanCassiere Do you mean that currently there is no way by `@tanstack/react-router` to bring a `data` object when `navigate`(by `useNavigate`) from one route to another?
